### ISSUE TITLE
[BUG] `--provenance` create provenance even if the publication fails

### ISSUE THREAD

**Issue by ericcornelissen** [npm/cli#7654]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

When trying to publish with `npm publish --provenance` and the publish fails, provenance is generated and published nonetheless. I'm unsure of all conditions under which this happens, but it definitely happens when trying to publish a version that has already been published.

The case where publication fails because the package is already published is particularly problematic because the result is that there's conflicting provenance about the version of a package out there (even if not directly linked in the registry).

### Expected Behavior

I would not expect provenance to be generated if nothing is published to the registry. A simple solution would be to publish the provenance only after the registry accepted the publication, but I guess this is problematic for two reasons: 1) I guess the registry should be informed explicitly about the existence of the provenance, 2) if the provenance generation fails it's hard to recover, which may lead to the package being published without provenance which is also not desirable.

### Steps To Reproduce

1. In GitHub Actions or GitLab CI/CD.
2. With the source code for an package at a version that has already been published.
3. Run `npm publish --provenance`.
4. See that 1) provenance is published while 2) the publish is rejected by the registry.
5. [example](https://gitlab.com/ericcornelissen/pp-test-kit/-/jobs/7338440013#L55)

### Environment

- Continuous integration
  - gitlab-runner 17.0.0~pre.88.g761ae5dd
  - [Node.js image `sha256:4d7bcfe46effa474ca1be502a1d97878bf972ac6507eda2cf93d57c09a9400d0`](https://registry.hub.docker.com/layers/library/node/latest/images/sha256-4d7bcfe46effa474ca1be502a1d97878bf972ac6507eda2cf93d57c09a9400d0)
- npm: 10.8.1
- Node.js: 22.4.1
- OS Name: bookworm
- System Model Name: _unknown_
- npm config:
```ini
; "user" config from ~/.npmrc

; n/a

; "project" config from .../pp-test-kit/.npmrc

audit = false 
ignore-scripts = true 
lockfile-version = "3" 
save-exact = true 
save-prefix = "" 
```

**Comment by bdehamer** [comment]
This is expected behavior for pretty much the reasons you outlined:

> I guess this is problematic for two reasons: 1) I guess the registry should be informed explicitly about the existence of the provenance, 2) if the provenance generation fails it's hard to recover, which may lead to the package being published without provenance which is also not desirable.

The provenance attestation is generated before the package is published -- this allows the attestation to be sent to the registry alongside the package it references. If the registry can't verify the supplied attestation the package will be rejected.

It's true that the attestation signature is published to the Rekor transparency log as part of the attestation process and this record WILL remain even if the package publication ultimately fails. This may result in a scenario where there are multiple entries in the transparency log that all refer to the same package but this doesn't really matter. The source of record for a package's attestation is the registry itself NOT the transparency log. Furthermore, the package's attestation will always have a reference to the correct transparency log entry.

**Comment by ericcornelissen** [comment]
> This is expected behavior for pretty much the reasons you outlined:
> 
> > I guess this is problematic for two reasons: 1) I guess the registry should be informed explicitly about the existence of the provenance, 2) if the provenance generation fails it's hard to recover, which may lead to the package being published without provenance which is also not desirable.
> 
> The provenance attestation is generated before the package is published -- this allows the attestation to be sent to the registry alongside the package it references. If the registry can't verify the supplied attestation the package will be rejected.

Fair enough. Thanks for confirming and expanding.

Is the source code, or conditions in plain English, for rejecting the package available somewhere?


> It's true that the attestation signature is published to the Rekor transparency log as part of the attestation process and this record WILL remain even if the package publication ultimately fails. This may result in a scenario where there are multiple entries in the transparency log that all refer to the same package but this doesn't really matter. The source of record for a package's attestation is the registry itself NOT the transparency log. Furthermore, the package's attestation will always have a reference to the correct transparency log entry.

I see your reasoning and that's enough of an answer to me - in that case I'll close this as is.
